library/pk_wrap.c

```diff
@@ -53,7 +53,23 @@ static int rsa_can_do(mbedtls_pk_type_t type)
 static size_t rsa_get_bitlen(const void *ctx)
 {
     const mbedtls_rsa_context *rsa = (const mbedtls_rsa_context *) ctx;
-    return 8 * mbedtls_rsa_get_len(rsa);
+    /* Unfortunately, the rsa.h interface does not have a direct way
+     * to access the bit-length that works with MBEDTLS_RSA_ALT.
+     * So we have to do a little work here.
+     */
+    mbedtls_mpi N;
+    mbedtls_mpi_init(&N);
+    int ret = mbedtls_rsa_export(rsa, &N, NULL, NULL, NULL, NULL);
+    /* If the export fails for some reason (e.g. the RSA_ALT implementation
+     * does not support export, or there is not enough memory),
+     * we have no way of returning an error from this function.
+     * As a fallback, return the byte-length converted in bits, which is
+     * the correct  value if the modulus size is a multiple of 8 bits, which
+     * is very often the case in practice. */
+    size_t bitlen = (ret == 0 ? mbedtls_mpi_bitlen(&N) :
+                     8 * mbedtls_rsa_get_len(rsa));
+    mbedtls_mpi_free(&N);
+    return bitlen;
```

Comment:
Given how much work is required to match the behaviour in `development`, and given that this would change the result of calling `rsa_get_bitlen()` in an LTS branch, should we actually make this change in 2.28?

i.e. how bad is "different behaviour in 2.28 and 3.x" vs "different behaviour in 2.28.7 and 2.28.8"?

Reply:
I'm not completely sure. I made a backport because it's a bug fix, and we normally backport bug fixes. By definitions, bug fixes change the behavior (except bug fixes that remove possibilities for undefined behavior). The real question is, how bad is “different behavior between library and expectations” vs “different behavior between 2.28.7 and 2.28.8”?

I do see the argument that this is a bug that has been publicly known for a long time, well before 2.28 became an LTS branch, so people who really care have presumably implemented a workaround, and the fix is of little value for long-established users. However, note that many new users do use an LTS — for example, many Linux distributions still ship 2.28.

The only risk I see with respect to the change in behavior is if callers are doing `mbedtls_pk_get_bitlen(pk) / 8` and expecting to get the byte size. I think this risk is low given the existence of `mbedtls_pk_get_len(pk)`. So I don't think the behavior change is bad enough to warrant not fixing the bug.